crates/key-value-spin/src/store.rs

```diff
@@ -156,6 +157,140 @@ impl Store for SqliteStore {
                 .collect()
         })
     }
+
+    async fn get_many(&self, keys: Vec<String>) -> Result<Vec<Option<(String, Vec<u8>)>>, Error> {
+        task::block_in_place(|| {
+            let sql_value_keys: Vec<rusqlite::types::Value> =
+                keys.into_iter().map(rusqlite::types::Value::from).collect();
+            let ptr = Rc::new(sql_value_keys);
+            let row_iter: Vec<Result<(String, Vec<u8>), Error>> = self.connection
+                .lock()
+                .unwrap()
+                .prepare_cached("SELECT key, value FROM spin_key_value WHERE store=:name AND key IN rarray(:keys)")
+                .map_err(log_error)?
+                .query_map((":name", &self.name, ":keys", ptr), |row| {
+                    <(String, Vec<u8>)>::try_from(row)
+                })
+                .map_err(log_error)?
+                .map(|r: Result<(String, Vec<u8>), rusqlite::Error>| r.map_err(log_error))
+                .collect();
+
+            let mut keys_and_values: Vec<Option<(String, Vec<u8>)>> = Vec::new();
+            for row in row_iter {
+                let res = row.map_err(log_error)?;
+                keys_and_values.push(Some(res));
+            }
+            Ok(keys_and_values)
+        })
+    }
+
+    async fn set_many(&self, key_values: Vec<(String, Vec<u8>)>) -> Result<(), Error> {
+        task::block_in_place(|| {
+            let mut binding = self.connection.lock().unwrap();
+            let tx = binding.transaction().map_err(log_error)?;
+            for kv in key_values {
+                tx.prepare_cached(
+                    "INSERT INTO spin_key_value (store, key, value) VALUES ($1, $2, $3)
```

Comment:
You're right, this is pretty unlovely, but I have no other ideas.  SQLite does provide for bulk inserts, but I don't think they let you do an update where the row already existed.  Hopefully it will not be too bad because in-process...!

Reply:
thankfully single-transaction-many-inserts is equally performant to single-transaction-bulk-insert - and this is easier to read (SQLite bulk inserts require the data source to be a `SELECT`, which is... unintuitive), so I wouldn't personally be too worried about perf here